File: spring-boot-project/src/main/java/com/example/demo/Controller/ForumController.java
package com.example.demo.Controller;

import com.example.demo.entities.User;
import com.example.demo.entities.forum.Comment;
import com.example.demo.entities.forum.ForumThread;
import com.example.demo.entities.forum.Post;
import com.example.demo.repository.UserRepository;
import com.example.demo.service.ForumService;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/forum")
public class ForumController {

    @Autowired
    private ForumService forumService;

    @Autowired
    private UserRepository userRepository;


    @GetMapping("/threads")
    public List<ForumThread> getAllThreads() {
        return forumService.getAllThreads();
    }

    @GetMapping("/threads/{threadId}/posts")
    public List<Post> getPostsByThread(@PathVariable Long threadId) {
        return forumService.getPostsByThreadId(threadId);
    }

    @GetMapping("/posts/{postId}/comments")
    public List<Comment> getCommentsByPost(@PathVariable Long postId) {
        return forumService.getCommentsByPostId(postId);
    }



    @PostMapping("/threads")
    public ForumThread createThread(@RequestBody CreateThreadRequest request) {
        User user = userRepository.findById(request.userId)
            .orElseThrow(() -> new IllegalArgumentException("Utilisateur pas trouve"));
        System.out.println(user);
        return forumService.createThread(request.title, user);
    }

    @PostMapping("/threads/{threadId}/posts")
    public Post createPost(@PathVariable Long threadId,
                           @RequestBody CreatePostRequest request) {
        User user = userRepository.findById(request.userId)
            .orElseThrow(() -> new IllegalArgumentException("Utilisateur pas trouve"));
        return forumService.createPost(threadId, request.content, user);
    }

    @PostMapping("/posts/{postId}/comments")
    public Comment addComment(@PathVariable Long postId,
                              @RequestBody AddCommentRequest request) {
        User user = userRepository.findById(request.userId)
            .orElseThrow(() -> new IllegalArgumentException("Utilisateur pas trouve"));
        return forumService.addComment(postId, request.text, user);
    }


}


class CreateThreadRequest {
    public String title;
    public int userId;
}

class CreatePostRequest {
    public String content;
    public int userId;
}

class AddCommentRequest {
    public String text;
    public int userId;
}

File: laravel-project/app/Http/Controllers/ForumController.php
<?php

namespace App\Http\Controllers;

use App\Services\ForumService;
use App\Models\Utilisateur;
use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;

class ForumController extends Controller
{
    protected ForumService $forumService;

    public function __construct(ForumService $forumService)
    {
        $this->forumService = $forumService;
    }

    /**
     * taamel thread jdid .
     */
    public function createThread(Request $request): JsonResponse
    {
        $request->validate([
            'title' => 'required|string',
            'author_id' => 'required|exists:utilisateur,id',
        ]);

        $user = Utilisateur::findOrFail($request->author_id);
        $thread = $this->forumService->createThread($request->title, $user);

        return response()->json($thread, 201);
    }

    /**
     * tejbed threads ml base
     */
    public function getAllThreads(): JsonResponse
    {
        $threads = $this->forumService->getAllThreads();
        return response()->json($threads);
    }

    /**
     * traja3 thread bel id .
     */
    public function getPostsByThreadId(int $threadId): JsonResponse
    {
        $posts = $this->forumService->getPostsByThreadId($threadId);
        return response()->json($posts);
    }

    /**
     *tamel post fel thread
     */
    public function createPost(int $threadId, Request $request): JsonResponse
    {
        $request->validate([
            'content' => 'required|string',
            'author_id' => 'required|exists:utilisateur,id',
        ]);

        $user = Utilisateur::findOrFail($request->author_id);
        $post = $this->forumService->createPost($threadId, $request->content, $user);

        return response()->json($post, 201);
    }

    /**
     * tzid post fel thread 
     */
    public function addComment(int $postId, Request $request): JsonResponse
    {
        $validated = $request->validate([
            'text' => 'required|string',
            'author_id' => 'required|exists:utilisateur,id',
        ]);

        $user = Utilisateur::findOrFail($validated['author_id']);
        $comment = $this->forumService->addComment($postId, $validated['text'], $user);

        return response()->json($comment, 201);
    }

    /**
     * Get all comments by post ID.
     */
    public function getCommentsByPostId(int $postId): JsonResponse
    {
        $comments = $this->forumService->getCommentsByPostId($postId);
        return response()->json($comments);
    }
}
